Validation/RecoParticleFlow/plugins/PFCaloGPUComparisonTask.cc

```diff
@@ -0,0 +1,170 @@
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+#include "DataFormats/ParticleFlowReco/interface/PFBlock.h"
+#include "DataFormats/ParticleFlowReco/interface/PFBlockElementCluster.h"
+#include "DataFormats/ParticleFlowReco/interface/PFBlockElementTrack.h"
+#include "DataFormats/ParticleFlowReco/interface/PFCluster.h"
+#include "DataFormats/ParticleFlowReco/interface/PFClusterFwd.h"
+#include "DataFormats/ParticleFlowReco/interface/PFLayer.h"
+#include "DataFormats/ParticleFlowReco/interface/PFRecHit.h"
+#include "DataFormats/ParticleFlowReco/interface/PFRecHitFraction.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/PluginManager/interface/ModuleDef.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#ifdef PFLOW_DEBUG
+#define LOGVERB(x) edm::LogVerbatim(x)
+#else
+#define LOGVERB(x) LogTrace(x)
+#endif
+
+class PFCaloGPUComparisonTask : public DQMEDAnalyzer {
+public:
+  PFCaloGPUComparisonTask(edm::ParameterSet const& conf);
+  ~PFCaloGPUComparisonTask() override = default;
+  void analyze(edm::Event const& e, edm::EventSetup const& c) override;
+  void bookHistograms(DQMStore::IBooker&, edm::Run const&, edm::EventSetup const&) override;
+
+private:
+  edm::EDGetTokenT<reco::PFClusterCollection> pfClusterTok_ref_;
+  edm::EDGetTokenT<reco::PFClusterCollection> pfClusterTok_target_;
+
+  MonitorElement* pfCluster_Multiplicity_GPUvsCPU_;
+  MonitorElement* pfCluster_Energy_GPUvsCPU_;
+  MonitorElement* pfCluster_RecHitMultiplicity_GPUvsCPU_;
+  MonitorElement* pfCluster_Layer_GPUvsCPU_;
+  MonitorElement* pfCluster_Depth_GPUvsCPU_;
+  MonitorElement* pfCluster_Eta_GPUvsCPU_;
+  MonitorElement* pfCluster_Phi_GPUvsCPU_;
+
+  std::string pfCaloGPUCompDir;
+};
+
+PFCaloGPUComparisonTask::PFCaloGPUComparisonTask(const edm::ParameterSet& conf)
+    : pfClusterTok_ref_{consumes<reco::PFClusterCollection>(
+          conf.getUntrackedParameter<edm::InputTag>("pfClusterToken_ref"))},
+      pfClusterTok_target_{
+          consumes<reco::PFClusterCollection>(conf.getUntrackedParameter<edm::InputTag>("pfClusterToken_target"))},
+      pfCaloGPUCompDir{conf.getUntrackedParameter<std::string>("pfCaloGPUCompDir")} {}
+
+void PFCaloGPUComparisonTask::bookHistograms(DQMStore::IBooker& ibooker,
+                                             edm::Run const& irun,
+                                             edm::EventSetup const& isetup) {
+  const char* histo;
+
+  ibooker.setCurrentFolder("ParticleFlow/" + pfCaloGPUCompDir);
+
+  histo = "pfCluster_Multiplicity_GPUvsCPU_";
```

Comment:
just curious, why do you call the histogram `"pfCluster_Multiplicity_GPUvsCPU_"` instead of simply `"pfCluster_Multiplicity_GPUvsCPU"` ?

Reply:
I can't recall particularly why we have the trailing underscore, but I don't believe it is necessary so I will remove it.